### PR TITLE
fix!: correctly handle nulls in btree and bitmap indices

### PR DESCRIPTION
**Pull request by westonpace**
There were a few issues with our null handling in scalar indices.

First, it appears I assumed earlier that `X < NULL` and `X > NULL` would always be false.  However, in `arrow-rs` the ordering considers `NULL` to be "the smallest value" and so `X < NULL` always evaluated to true.  This required some changes to the logic in the btree and bitmap indices.

Second, the btree index was still using the v1 file format because it relied on the page size to keep track of the index's batch size.  I've instead made the batch size a configurable property (configurable in code, not configurable by users) and made it so that btree can use the v2 file format.

Finally, related to the above, I changed it so we now write v2 files for all scalar indices, even if the dataset is a v1 dataset.  I think that's a reasonable decision at this point.

The logic to fallback and read the old v1 files was already in place (I believe @BubbleCal added it back when working on inverted index) but I added a migration test just to be sure we weren't breaking our btree / bitmap support.

Users with existing bitmap indices will get the new correct behavior without any changes.
Users with existing btree indices will get some of the new correct behavior but will need to retrain their indices to get all of the correct behavior. 

BREAKING CHANGE: Bitmap and btree indices will no longer be readable by older versions of Lance.  This is not a "backwards compatibility change" (no APIs or code will stop working) but rather a "forwards compatibility change" (you need to be careful in a multi-verison deployment or if you roll back)

### CODE REVIEW

File: rust/lance-index/src/scalar/btree.rs [hunk 1]
```diff
@@ -669,13 +674,13 @@ impl BTreeLookup {
 /// Note: this is very similar to the IVF index except we store the IVF part in a btree
 /// for faster lookup
 #[derive(Clone, Debug, DeepSizeOf)]
-pub struct BTreeIndex {
+pub struct BTreeIndex<const BATCH_SIZE: u64 = DEFAULT_BTREE_BATCH_SIZE> {
```
Comment (wjones127) [inline rust/lance-index/src/scalar/btree.rs:677]: What's the benefit of making `BATCH_SIZE` a compile-time constant?

**Review by wjones127 — APPROVED** [review]
Looks good. Although I think we might consider this a breaking change, since IIUC older versions won't be able to read the new scalar indices, right? Or will they?

File: rust/lance-index/src/scalar/btree.rs (hunk 1 above)
Comment (westonpace) [reply rust/lance-index/src/scalar/btree.rs:677]: Eh, it was less that I wanted to make it a compile time constant and more that I didn't have a good way of passing the batch size to `BTreeIndex::load` (a static method).  However, I realize now I can put it in the file schema.  Let me do that real quick.

**Comment by westonpace** [comment]
> Although I think we might consider this a breaking change, since IIUC older versions won't be able to read the new scalar indices, right? Or will they?

They will not be able to.  They would get a runtime error about "legacy method called on v2 file".  However, that's more of a forward compatibility concern right?  This is _technically_ not a breaking change according to [semver](https://github.com/semver/semver/issues/275) but I'm fine considering it as one.  Otherwise A/B style upgades will break  I will update the title.

**Review by wjones127 — APPROVED** [review]

**Pull request merged**